src/test/ui/consts/miri_unleashed/mutable_const.rs

```diff
@@ -0,0 +1,20 @@
+// compile-flags: -Zunleash-the-miri-inside-of-you
+
+#![feature(const_raw_ptr_deref)]
+#![deny(const_err)]
+
+use std::cell::UnsafeCell;
+
+// make sure we do not just intern this as mutable
+const MUTABLE_BEHIND_RAW: *mut i32 = &UnsafeCell::new(42) as *const _ as *mut _;
+
+const MUTATING_BEHIND_RAW: () = {
+    // Test that `MUTABLE_BEHIND_RAW` is actually immutable, by doing this at const time.
+    unsafe {
+        *MUTABLE_BEHIND_RAW = 99 //~ WARN skipping const checks
+        //~^ ERROR any use of this value will cause an error
+        //~^^ tried to modify constant memory
```

Comment:
I guess the interesting question this test exposes is: if there is a way to sneak by the "constants are immutable" check, the way this code here does, what is the desired behavior? Should it be UB to mutate that memory (that's what this PR does) or just observably different from when the constant got inlined everywhere (which used to be the case before)?

I wish we could just error when we find untyped pointers in a constant. But, well, this compiles on stable:
```rust
use std::ptr::NonNull;

union Transmute {
    reference: &'static u8,
    nonnull: NonNull<u8>,
}

const NON_NULL_PTR: NonNull<u8> = unsafe { Transmute { reference: &0 }.nonnull };
```
The allocation here probably arises from a `force_allocation`. I wonder if there are circumstances under which that could make the new allocation read-only? Probably not, before even for immutable bindings we often have to `force_allocation` before initializing, as we can only initialize the value in memory.

Reply:
There's no need for unsafe in order to get untyped pointers in Constants, just initializing a Union with a pointer and having a constant of Union type will get you that.

Reply:
Ah, good point:
```rust
union U {
    reference: &'static u8,
    int: usize,
}

const NON_NULL_PTR: U = U { reference: &0 };
```

Reply:
> Should it be UB to mutate that memory (that's what this PR does) or just observably different from when the constant got inlined everywhere (which used to be the case before)?

In my experience everyone has argued for that to be UB. In fact, once we unify promoteds between functions and constants to always create an anonymous immutable static, that will be the behaviour anyway

Reply:
Given that this was UB before your interning refactoring, I think I am fine making it UB again.